File: yonote_client.py
import requests


class YonoteClient:
    """Client for Yonote API (Outline-compatible)."""

    def __init__(self, api_token, base_url="https://app.yonote.ru/api"):
        self.api_token = api_token
        self.base_url = base_url.rstrip("/")
        self.headers = {
            "Authorization": f"Bearer {api_token}",
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        self._workspace_url = None

    def _post(self, endpoint, data=None):
        """Make a POST request to the Yonote API."""
        url = f"{self.base_url}/{endpoint}"
        response = requests.post(url, headers=self.headers, json=data or {}, timeout=30)
        response.raise_for_status()
        return response.json()

    # --- Auth ---

    def auth_info(self):
        """Get current user and team info."""
        return self._post("auth.info")

    def get_workspace_url(self):
        """Get the workspace base URL (e.g. https://remake.yonote.ru). Cached after first call."""
        if self._workspace_url is None:
            result = self.auth_info()
            self._workspace_url = result.get("data", {}).get("team", {}).get("url", "")
        return self._workspace_url

    def full_url(self, relative_url):
        """Convert a relative document URL to a full URL."""
        if not relative_url:
            return ""
        if relative_url.startswith("http"):
            return relative_url
        workspace = self.get_workspace_url()
        return f"{workspace}{relative_url}"

    # --- Collections ---

    def collections_list(self, limit=25, offset=0):
        """List all collections."""
        return self._post("collections.list", {"limit": limit, "offset": offset})

    def collection_info(self, collection_id):
        """Get collection details."""
        return self._post("collections.info", {"id": collection_id})

    def collection_documents(self, collection_id):
        """List documents in a collection."""
        return self._post("collections.documents", {"id": collection_id})

    # --- Documents ---

    def documents_list(self, collection_id=None, parent_document_id=None, limit=25, offset=0):
        """List documents, optionally filtered by collection or parent document."""
        data = {"limit": limit, "offset": offset}
        if collection_id:
            data["collectionId"] = collection_id
        if parent_document_id:
            data["parentDocumentId"] = parent_document_id
        return self._post("documents.list", data)

    def document_info(self, document_id):
        """Get document details."""
        return self._post("documents.info", {"id": document_id})

    def documents_search(self, query, collection_id=None, limit=25, offset=0):
        """Full-text search across documents."""
        data = {"query": query, "limit": limit, "offset": offset}
        if collection_id:
            data["collectionId"] = collection_id
        return self._post("documents.search", data)

    def document_create(self, title, text="", collection_id=None, parent_document_id=None, publish=True):
        """Create a new document."""
        data = {"title": title, "text": text, "publish": publish}
        if collection_id:
            data["collectionId"] = collection_id
        if parent_document_id:
            data["parentDocumentId"] = parent_document_id
        return self._post("documents.create", data)

    def document_update(self, document_id, title=None, text=None, append=False):
        """Update an existing document. Use append=True to add text to end instead of replacing."""
        data = {"id": document_id}
        if title is not None:
            data["title"] = title
        if text is not None:
            data["text"] = text
        if append:
            data["append"] = True
        return self._post("documents.update", data)

    def document_delete(self, document_id, permanent=False):
        """Delete a document."""
        return self._post("documents.delete", {"id": document_id, "permanent": permanent})

    def document_move(self, document_id, collection_id=None, parent_document_id=None):
        """Move document to different collection or parent."""
        data = {"id": document_id}
        if collection_id:
            data["collectionId"] = collection_id
        if parent_document_id:
            data["parentDocumentId"] = parent_document_id
        return self._post("documents.move", data)

    def document_archive(self, document_id):
        """Archive a document (soft delete)."""
        return self._post("documents.archive", {"id": document_id})

    def document_restore(self, document_id):
        """Restore an archived/deleted document."""
        return self._post("documents.restore", {"id": document_id})

    def documents_drafts(self, limit=25, offset=0):
        """List user's draft documents."""
        return self._post("documents.drafts", {"limit": limit, "offset": offset})

    def documents_viewed(self, limit=25, offset=0):
        """List recently viewed documents."""
        return self._post("documents.viewed", {"limit": limit, "offset": offset})

    # --- Collections (extended) ---

    def collection_create(self, name, description=""):
        """Create a new collection."""
        data = {"name": name}
        if description:
            data["description"] = description
        return self._post("collections.create", data)

    def collection_delete(self, collection_id):
        """Delete a collection."""
        return self._post("collections.delete", {"id": collection_id})

File: js/yonote-client.js
/**
 * Yonote API client — port of yonote_client.py.
 * All requests go through the PHP proxy to avoid CORS issues.
 */
export class YonoteClient {
    constructor(apiToken, baseUrl = 'https://app.yonote.ru/api') {
        this.apiToken = apiToken;
        this.baseUrl = baseUrl.replace(/\/+$/, '');
        this._workspaceUrl = null;
    }

    async _post(endpoint, data = {}) {
        const url = `${this.baseUrl}/${endpoint}`;
        const resp = await fetch('/api/proxy.php', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                url,
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.apiToken}`,
                    'Content-Type': 'application/json',
                    'Accept': 'application/json',
                },
                body: data,
            }),
        });
        if (!resp.ok) {
            const text = await resp.text();
            throw new Error(`API error ${resp.status}: ${text}`);
        }
        return resp.json();
    }

    // --- Auth ---

    async authInfo() {
        return this._post('auth.info');
    }

    async getWorkspaceUrl() {
        if (this._workspaceUrl === null) {
            const result = await this.authInfo();
            this._workspaceUrl = result?.data?.team?.url || '';
        }
        return this._workspaceUrl;
    }

    fullUrl(relativeUrl) {
        if (!relativeUrl) return '';
        if (relativeUrl.startsWith('http')) return relativeUrl;
        // Use cached workspace URL (must call getWorkspaceUrl first)
        return `${this._workspaceUrl || ''}${relativeUrl}`;
    }

    getAttachmentUrl(attachmentId) {
        return `${this.baseUrl}/attachments.redirect?id=${attachmentId}`;
    }

    // --- Collections ---

    async collectionsList(limit = 25, offset = 0) {
        return this._post('collections.list', { limit, offset });
    }

    async collectionInfo(collectionId) {
        return this._post('collections.info', { id: collectionId });
    }

    async collectionDocuments(collectionId) {
        return this._post('collections.documents', { id: collectionId });
    }

    async collectionCreate(name, description = '') {
        const data = { name };
        if (description) data.description = description;
        return this._post('collections.create', data);
    }

    async collectionDelete(collectionId) {
        return this._post('collections.delete', { id: collectionId });
    }

    // --- Documents ---

    async documentsList(collectionId = null, parentDocumentId = null, limit = 25, offset = 0) {
        const data = { limit, offset };
        if (collectionId) data.collectionId = collectionId;
        if (parentDocumentId) data.parentDocumentId = parentDocumentId;
        return this._post('documents.list', data);
    }

    async documentInfo(documentId) {
        return this._post('documents.info', { id: documentId });
    }

    async documentsSearch(query, collectionId = null, limit = 25, offset = 0) {
        const data = { query, limit, offset };
        if (collectionId) data.collectionId = collectionId;
        return this._post('documents.search', data);
    }

    async documentCreate(title, text = '', collectionId = null, parentDocumentId = null, publish = true) {
        const data = { title, text, publish };
        if (collectionId) data.collectionId = collectionId;
        if (parentDocumentId) data.parentDocumentId = parentDocumentId;
        return this._post('documents.create', data);
    }

    async documentUpdate(documentId, title = null, text = null, append = false) {
        const data = { id: documentId };
        if (title !== null) data.title = title;
        if (text !== null) data.text = text;
        if (append) data.append = true;
        return this._post('documents.update', data);
    }

    async documentDuplicate(documentId, title = null, publish = true, recursive = false) {
        const data = { id: documentId, publish };
        if (title) data.title = title;
        if (recursive) data.recursive = true;
        return this._post('documents.duplicate', data);
    }

    async documentDelete(documentId, permanent = false) {
        return this._post('documents.delete', { id: documentId, permanent });
    }

    async documentMove(documentId, collectionId = null, parentDocumentId = null) {
        const data = { id: documentId };
        if (collectionId) data.collectionId = collectionId;
        if (parentDocumentId) data.parentDocumentId = parentDocumentId;
        return this._post('documents.move', data);
    }

    async documentArchive(documentId) {
        return this._post('documents.archive', { id: documentId });
    }

    async documentRestore(documentId) {
        return this._post('documents.restore', { id: documentId });
    }

    async documentsDrafts(limit = 25, offset = 0) {
        return this._post('documents.drafts', { limit, offset });
    }

    async documentsViewed(limit = 25, offset = 0) {
        return this._post('documents.viewed', { limit, offset });
    }

    // --- Export ---

    async documentExportMarkdown(documentId, timeout = 30000) {
        // Step 1: Start export
        const result = await this._post('documents.export', { id: documentId });
        const foId = result.data.fileOperation.id;

        // Step 2: Poll until complete
        const deadline = Date.now() + timeout;
        while (Date.now() < deadline) {
            await new Promise(r => setTimeout(r, 500));
            const status = await this._post('fileOperations.info', { id: foId });
            const state = status?.data?.state;
            if (state === 'complete') break;
            if (state === 'error') throw new Error('Export failed');
        }
        if (Date.now() >= deadline) {
            throw new Error(`Export not completed within ${timeout}ms`);
        }

        // Step 3: Get redirect URL (noRedirect mode)
        const redirectResp = await fetch('/api/proxy.php', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                url: `${this.baseUrl}/fileOperations.redirect`,
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.apiToken}`,
                    'Content-Type': 'application/json',
                },
                body: { id: foId },
                noRedirect: true,
            }),
        });
        const redirectData = await redirectResp.json();
        const redirectUrl = redirectData.location;
        if (!redirectUrl) throw new Error('No redirect URL from export');

        // Step 4: Download content
        const downloadResp = await fetch('/api/proxy.php', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                url: redirectUrl,
                method: 'GET',
                headers: {},
            }),
        });
        const downloadData = await downloadResp.json();
        return downloadData._rawText || '';
    }

    // --- Attachments ---

    async attachmentsList(documentId) {
        return this._post('attachments.list', { documentId });
    }
}
